Rewrite `use()` docs (#8305)
reactjs/reactjs.org: src/content/reference/react/use.md

Reviewer: Agree! To add to this, the "RSC blocks rendering" framing can read as if it blocks the entire app, when  that's not really the case. Maybe we can mention that if we move the async work to a child async Server Component wrapped in `<Suspense>`, it enables SSR streaming and avoids blocking.

I think this Deep Dive (or somewhere else in these use() docs) would benefit from mentioning the async SC + Suspense pattern and clarifying when each approach is preferable. Also, it's worth noting that neither the Suspense docs nor the RSC docs currently cover this "stremaing SSR" pattern either. Could be a good follow-up to add streaming examples there as well :)

Reviewer 2: TIL.  I definitely believed it was the `await` itself that was holding up the initial SSR response and that the lack of `await` (regardless of `Suspense` usage) is what allows us to get around this.  Surprised to find that `use()` also blocks the initial SSR unless you use `Suspense`.

Reviewer 3: This nerd sniped me into creating a quick demo comparison of async and use. The bottom line is that Suspense + use/async is magical when you isolate slow data within Suspense boundaries.

demo link: https://streaming-test-three.vercel.app/comparison 
repo: https://github.com/maxwellCohen/streaming-test/

Reviewer 4: I tried to adress all this here! Let me know if its more clear. https://github.com/reactjs/react.dev/pull/8305/changes/ceab6dadbfd754ef3c83a02b8a3cd182d501ed65

Reviewer 3: [nit] use() in React 19 can only be used for promises or context. The "like" seems to imply you can use it with other data resources.    

So something like this would be more accurate:
```
`use` is a React API that lets you read the value of a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) resource or [context](/learn/passing-data-deeply-with-context) resource.
```

Ther e
https://github.com/react/react/blob/b91823e21434ac665450e67ddc6f816710255938/packages/react-server/src/ReactFizzHooks.js#L748-L763

Reviewer 4: https://github.com/rickhanlonii/react.dev/commit/5d3e81af Adressed! good catch. Afaik in the future there might be more "resources", probably why it was framed like that!

<Intro>

> `use` is a React API that lets you read the value of a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) or [context](/learn/passing-data-deeply-with-context) resource.

Author: ```suggestion
`use` is a React API that lets you read the value of a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) or [context](/learn/passing-data-deeply-with-context).
```

```js
import { use } from 'react';

function MessageComponent({ messagePromise }) {
  const message = use(messagePromise);
  // ...
```

If the component that calls `use` is wrapped in a [`Suspense`](/reference/react/Suspense) boundary, the fallback will be displayed while the Promise is pending. Once the Promise is resolved, the Suspense fallback is replaced by the rendered components using the data returned by `use`. If the Promise is rejected, the fallback of the nearest [Error Boundary](/reference/react/Component#catching-rendering-errors-with-an-error-boundary) will be displayed.

[See more examples below.](#usage-promises)

#### Parameters {/*promise-parameters*/}

* `promise`: A [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) whose resolved value you want to read. The Promise must be [cached](#caching-promises-for-client-components) so that the same instance is reused across re-renders.

#### Returns {/*promise-returns*/}

The resolved value of the Promise.

#### Caveats {/*promise-caveats*/}

* `use` must be called inside a Component or a Hook.

> * Despite its name, `use` is not a Hook. Unlike Hooks, it can be called inside loops and conditional statements like `if`.

Reviewer 2: **nit**:  Small asymmetry now is that this is in the caveats section here, but for use(context) it's not in the caveats section, but rather the 2nd sentence in the main body. Really small, just mentioning to be exhaustive.

Reviewer 4: Nice catch, fixed! https://github.com/reactjs/react.dev/pull/8305/changes/c46d999a1247842c2d1a581dd586159be00a8df6

* `use` cannot be called inside a try-catch block. Instead, wrap your component in an [Error Boundary](#displaying-an-error-with-an-error-boundary) to catch the error and display a fallback.
* Promises passed to `use` must be cached so the same Promise instance is reused across re-renders. [See caching Promises below.](#caching-promises-for-client-components)

> * When passing a Promise from a Server Component to a Client Component, its resolved value must be [serializable](/reference/rsc/use-client#serializable-types). Data types like functions aren't serializable and cannot be the resolved value of such a Promise.

Reviewer 2: > Data types like functions aren't serializable and cannot be the resolved value of such a Promise.

Doesn't seem fully accurate as client references can be passed.  My take would be to just leave this sentence out and let the linked page take care of that nuance.

Reviewer 4: Done! https://github.com/reactjs/react.dev/pull/8305/changes/c46d999a1247842c2d1a581dd586159be00a8df6 thanks

If the component that calls `use` is wrapped in a [`Suspense`](/reference/react/Suspense) boundary, the fallback will be displayed while the Promise is pending. Once the Promise is resolved, the Suspense fallback is replaced by the rendered components using the data returned by `use`. If the Promise is rejected, the fallback of the nearest [Error Boundary](/reference/react/Component#catching-rendering-errors-with-an-error-boundary) will be displayed.

#### Returns {/*returns*/}
[See more examples below.](#usage-promises)

The `use` API returns the value that was read from the resource like the resolved value of a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) or [context](/learn/passing-data-deeply-with-context).
#### Parameters {/*promise-parameters*/}

#### Caveats {/*caveats*/}
* `promise`: A [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) whose resolved value you want to read. The Promise must be [cached](#caching-promises-for-client-components) so that the same instance is reused across re-renders.

* The `use` API must be called inside a Component or a Hook.

> * When fetching data in a [Server Component](/reference/rsc/server-components), prefer `async` and `await` over `use`. `async` and `await` pick up rendering from the point where `await` was invoked, whereas `use` re-renders the component after the data is resolved.

Author: It's covered in "Should I resolve a Promise in a Server or Client Component" below.

> * `use` must be called inside a Component or a Hook.

Reviewer 4: Fixed! https://github.com/reactjs/react.dev/pull/8305/changes/ceab6dadbfd754ef3c83a02b8a3cd182d501ed65

> * When passing a Promise from a Server Component to a Client Component, its resolved value must be serializable to pass between server and client. Data types like functions aren't serializable and cannot be the resolved value of such a Promise.

Reviewer: > [...] to pass between server and client.

nit: for me this phrase feels redundant here, and it reads a bit weird when I first read it. It also subtly shifts from React terminology to infra terminology, which could be confusing since RSC don't necessarily imply a physical server. I would just drop that phrase:

```suggestion
* When passing a Promise from a Server Component to a Client Component, its resolved value must be serializable. Data types like functions aren't serializable and cannot be the resolved value of such a Promise.
```

> * When passing a Promise from a Server Component to a Client Component, its resolved value must be serializable to pass between server and client. Data types like functions aren't serializable and cannot be the resolved value of such a Promise.

Reviewer 4: I think this should be adressed here! https://github.com/reactjs/react.dev/pull/8305/changes/ceab6dadbfd754ef3c83a02b8a3cd182d501ed65
have a look

```

```js src/data.js hidden
// Note: the way you would do data fetching depends on
// the framework that you use together with Suspense.
// Normally, the caching logic would be inside a framework.

let cache = new Map();

export function fetchData(url) {
  if (!cache.has(url)) {
    cache.set(url, getData(url));
  }
  return cache.get(url);
}

async function getData(url) {
  if (url === '/albums') {
    return await getAlbums();
  } else {
    throw Error('Not implemented');
  }
}

async function getAlbums() {
  // Add a fake delay to make waiting noticeable.
  await new Promise(resolve => {
    setTimeout(resolve, 1000);
  });

  return [{
    id: 13,
    title: 'Let It Be',
    year: 1970
  }, {
    id: 12,
    title: 'Abbey Road',
    year: 1969
  }, {
    id: 11,
    title: 'Yellow Submarine',
    year: 1969
  }, {
    id: 10,
    title: 'The Beatles',
    year: 1968
  }];
}
```

```json package.json hidden
{
  "dependencies": {
    "react": "19.0.0",
    "react-dom": "19.0.0",
    "react-scripts": "^5.0.0",
    "react-error-boundary": "4.0.3"
  },
  "main": "/index.js"
}
```

</Sandpack>

<Solution />

#### Fetching data with `useEffect` {/*fetching-data-with-useeffect*/}

> Without `use`, a common approach is to fetch data in an Effect and update state when the data arrives. This requires managing loading and error states manually, and the component renders empty on first paint before the Effect fires.

Reviewer 5: would it make sense to mention this isn't recommended and you should use `use`. Maybe thats obvious

Reviewer 4: Added a link to useEffect docs here, is that helpful? https://github.com/reactjs/react.dev/pull/8305/changes/ceab6dadbfd754ef3c83a02b8a3cd182d501ed65

export default function App() {
  const [albums, setAlbums] = useState(null);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    fetchAlbums()
      .then(data => {
        setAlbums(data);
        setIsLoading(false);
      })
      .catch(err => {
        setError(err);
        setIsLoading(false);
      });
  }, []);

  if (isLoading) {
    return <h2>Loading...</h2>;
  }

  if (error) {
    return <p>Error: {error.message}</p>;
  }

  return (
    <ul>
      {albums.map(album => (
        <li key={album.id}>
          {album.title} ({album.year})
        </li>
      ))}
    </ul>
  );
}
```

```js src/data.js hidden
export async function fetchAlbums() {
  // Add a fake delay to make waiting noticeable.
  await new Promise(resolve => {
    setTimeout(resolve, 1000);
  });

  return [{
    id: 13,
    title: 'Let It Be',
    year: 1970
  }, {
    id: 12,
    title: 'Abbey Road',
    year: 1969
  }, {
    id: 11,
    title: 'Yellow Submarine',
    year: 1969
  }, {
    id: 10,
    title: 'The Beatles',
    year: 1968
  }];
}
```

</Sandpack>

<Solution />

</Recipes>

<Pitfall>

> ##### Promises passed to `use` must be cached {/*promises-must-cached*/}

Reviewer 5: I really like this section as I think its a common pitfall

Promises created inside a component are recreated on every render. This causes React to show the Suspense fallback repeatedly and prevents content from appearing. Instead, pass a Promise from a cache, a Suspense-enabled framework, or a Server Component.

```js
function Albums() {
  // 🔴 fetch creates a new Promise on every render.

>   const albums = use(fetch('/albums'));

Reviewer 4: Fixed! https://github.com/reactjs/react.dev/pull/8305/changes/ceab6dadbfd754ef3c83a02b8a3cd182d501ed65

  // ...
}
```

```js
// ✅ fetchData reads the promise from a cache.
const albums = use(fetchData('/albums'));
```

Ideally, Promises are created before rendering — such as in an event handler, a route loader, or a Server Component — and passed to the component that calls `use`. Fetching lazily in render delays network requests and can create waterfalls.

</Pitfall>

---

### Caching Promises for Client Components {/*caching-promises-for-client-components*/}

Promises passed to `use` in Client Components must be cached so the same Promise instance is reused across re-renders. If a new Promise is created directly in render, React will display the Suspense fallback on every re-render.

```js
// ✅ Cache the Promise so the same one is reused across renders
let cache = new Map();

export function fetchData(url) {
  if (!cache.has(url)) {

>     cache.set(url, getData(url));

Reviewer 2: I'm trying to reconcile this with the Rules of React.  My first thought is that this is "okay" because it falls under [lazy initialization](https://react.dev/reference/rules/components-and-hooks-must-be-pure#lazy-initialization).  Is that how you think about it?  

In the example given on that page, it says "Good: if it doesn't affect other components" and this will affect other components in some sense, but I guess it affects them in a way that doesn't really matter.

Reviewer 3: The way I think about `use` for promises is that it is like an useEffect, where you pass in a promise dependency. As the promise changes status, `use` will run various actions (throwing an error if not resolved) and rerender when resolved. Just like you cannot put a newly created function/object in an useEffect depency array, you cannot pass a new promise into `use`

I am very open to other ways of thinking about this.

Author: Yeah it's like a lazy init pattern. Technically, setting the initial state of a hook, or setting the initial value of a ref, or sync'ing with set state in render are all side effects. But they are all still pure. But ideally, you're prewarming the cache (like setting the promise in a route loader, or passing the promise from a server component).

  );
}

function Loading() {
  return <h2>Loading...</h2>;
}
```

```js src/data.js hidden
// Note: the way you would do data fetching depends on
// the framework that you use together with Suspense.
// Normally, the caching logic would be inside a framework.

let cache = new Map();

export function fetchData(url) {
  if (!cache.has(url)) {
    cache.set(url, getData(url));
  }
  return cache.get(url);
}

export function refetchData(url) {
  cache.delete(url);
  return fetchData(url);
}

async function getData(url) {
  if (url.startsWith('/the-beatles/albums')) {
    return await getAlbums();
  } else {
    throw Error('Not implemented');
  }
}

async function getAlbums() {
  // Add a fake delay to make waiting noticeable.
  await new Promise(resolve => {
    setTimeout(resolve, 1000);
  });

  return [{
    id: 13,
    title: 'Let It Be',
    year: 1970
  }, {
    id: 12,
    title: 'Abbey Road',
    year: 1969
  }, {
    id: 11,
    title: 'Yellow Submarine',
    year: 1969
  }, {
    id: 10,
    title: 'The Beatles',
    year: 1968
  }, {
    id: 9,
    title: 'Magical Mystery Tour',
    year: 1967
  }];
}
```

```css
button { margin-bottom: 10px; }
```

</Sandpack>

<Note>

> Frameworks that support Suspense typically provide their own caching and invalidation mechanisms. Building a custom cache like the one above is useful for understanding the pattern, but in practice you should use your framework's data fetching solution.

Reviewer 2: If you wanted to use React without a framework and just use `use()` either via an underlying library or building your own, would that also be considered valid usage?  I think I recall talks of trying to replace `useEffect` in react-query with use.

It's just framework made me think meta-framework as in only NextJS, RR7, Tanstack Start, etc.

Author: I don't really understand the framing of "valid usage" tbh. I'm not really here to be the React police and tell people what's valid and not valid to do. The note here is just saying if you're using a framework, favor their patterns. But you're free to create your own patterns or libraries, or to even ignore that advice and roll your own inside a framework.

// the framework that you use together with Suspense.
// Normally, the caching logic would be inside a framework.

let cache = new Map();

export function fetchData(url) {
  if (!cache.has(url)) {
    const promise = getData(url);
    // Set status fields so React can read the value
    // synchronously if the Promise resolves before
    // `use` is called (e.g. when preloading on hover).
    promise.status = 'pending';
    promise.then(
      value => {
        promise.status = 'fulfilled';
        promise.value = value;
      },
      reason => {
        promise.status = 'rejected';
        promise.reason = reason;
      },
    );
    cache.set(url, promise);
  }
  return cache.get(url);
}

async function getData(url) {
  if (url.startsWith('/the-beatles/albums')) {
    return await getAlbums('the-beatles');
  } else if (url.startsWith('/led-zeppelin/albums')) {
    return await getAlbums('led-zeppelin');
  } else if (url.startsWith('/pink-floyd/albums')) {
    return await getAlbums('pink-floyd');
  } else {
    throw Error('Not implemented');
  }
}

async function getAlbums(artistId) {
  // Add a fake delay to make waiting noticeable.
  await new Promise(resolve => {

>     setTimeout(resolve, 80);

Reviewer 4: fixed https://github.com/reactjs/react.dev/pull/8305/changes/ceab6dadbfd754ef3c83a02b8a3cd182d501ed65

</Sandpack>

---

## Usage (Promises) {/*usage-promises*/}

### Reading a Promise with `use` {/*reading-a-promise-with-use*/}

Call `use` with a Promise to read its resolved value. The component will [suspend](/reference/react/Suspense) while the Promise is pending.

```js [[1, 4, "use(albumsPromise)"]]
import { use } from 'react';

function Albums({ albumsPromise }) {
  const albums = use(albumsPromise);
  return (
    <ul>
      {albums.map(album => (
        <li key={album.id}>
          {album.title} ({album.year})
        </li>
      ))}
    </ul>
  );
}
```

> The component that calls <CodeStep step={1}>`use`</CodeStep> must be wrapped in a [`Suspense`](/reference/react/Suspense) boundary. While the Promise is pending, the Suspense fallback is displayed. Once the Promise resolves, React reads the value with `use` and replaces the fallback with the rendered component.

Reviewer 3: Do we need a deep dive into how `use(promise)` and `Suspense` work, or more details about how use under the hood throws an error, or a little more details like this: 

While the Promise is pending, `use` throws an internal error, and Suspense catches the error and displays a fallback. Once the Promise resolves, React replaces the fallback with the freshly rendered component, reading the resolved promise value with `use`.

Reviewer 4: Thanks for the suggestion! I think we can leave this one as-is. The internals (use throwing, Suspense catching) are already covered in the Troubleshooting section about try/catch, where it's more relevant to the reader's problem.

  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    fetchAlbums()
      .then(data => {
        setAlbums(data);
        setIsLoading(false);
      })
      .catch(err => {
        setError(err);
        setIsLoading(false);
      });
  }, []);

  if (isLoading) {
    return <h2>Loading...</h2>;
  }

  if (error) {
    return <p>Error: {error.message}</p>;
  }

  return (
    <ul>
      {albums.map(album => (
        <li key={album.id}>
          {album.title} ({album.year})
        </li>
      ))}
    </ul>
  );
}
```

```js src/data.js hidden
export async function fetchAlbums() {
  // Add a fake delay to make waiting noticeable.
  await new Promise(resolve => {
    setTimeout(resolve, 1000);
  });

  return [{
    id: 13,
    title: 'Let It Be',
    year: 1970
  }, {
    id: 12,
    title: 'Abbey Road',
    year: 1969
  }, {
    id: 11,
    title: 'Yellow Submarine',
    year: 1969
  }, {
    id: 10,
    title: 'The Beatles',
    year: 1968
  }];
}
```

</Sandpack>

<Solution />

</Recipes>

<Pitfall>

##### Promises passed to `use` must be cached {/*promises-must-cached*/}

> Promises created inside a component are recreated on every render. This causes React to show the Suspense fallback repeatedly and prevents content from appearing. Instead, pass a Promise from a cache, a Suspense-enabled framework, or a Server Component.

Reviewer 3: We might want to add more of an explanation that when components re-render, React will not preserve any previous state 

``` 
Promises created inside a component are recreated on every render because React does not preserve any state for renders that were suspended before they could mount for the first time. When the component has loaded, React will retry rendering the suspended tree from scratch. Recreating promises within the component causes React to repeatedly render the Suspense fallback, preventing content from appearing. Using internal component caching, such as useState factory functions, useMemo, or the React compiler, will cause promises to be recreated. Instead, pass a Promise from an external cache, a Suspense-enabled framework, or a Server Component.
```

Reviewer 4: Added a quick link! Hows that? https://github.com/reactjs/react.dev/pull/8305/changes/92208c21624e244d5608a27586def0b6c8cbe6a6

Author: Looks good for now, but note that this is subject to change in the future (no need to mention that though).

Common ways a Promise can be unintentionally recreated on every render:

```js
function Albums() {
  // 🔴 `fetch` creates a new Promise on every render.
  const albums = use(fetch('/albums'));

  // 🔴 Calling an async function creates a new Promise on every render.

>   const albums = use(getAlbums());

Author: @aurorascharff this is confusing because it's not clear how `getAlbums` and `fetchData` are different.

Reviewer 4: Hows this? https://github.com/reactjs/react.dev/pull/8305/commits/3cf1bf539b636a255e769341657b570ac9bda2cc

Author: Looks good!

</Sandpack>

### Reading a Promise from context {/*reading-a-promise-from-context*/}

You can pass a Promise through context to share data without prop drilling. Pass the Promise as the context value, then call `use(context)` to read the Promise and `use(promise)` to read its resolved value:

```js
import { use } from 'react';
import { UserContext } from './UserContext';

function Profile() {
  const userPromise = use(UserContext);
  const user = use(userPromise);
  return <h1>{user.name}</h1>;
}
```

Wrap the components that read the Promise in [`<Suspense>`](/reference/react/Suspense) so only that subtree suspends while the Promise is pending. See [Usage (Promises)](#usage-promises) below for more on reading Promises with `use`.

<Pitfall>

> Passing a Promise through context is convenient, but comes with trade-offs. Prefer passing the Promise as a [prop](#reading-a-promise-with-use) from a [Server Component](/reference/rsc/server-components) when possible.

Reviewer 2: Should we also mention reading directly from the cache (ie, the fetchData function using the cache) here as well (and possibly link to that section)? Another approach would be to mention that passing promises via context is more for server-initiated fetches.

In terms of deciding when to use each approach, I'd think client-initiated fetches should probably prefer reading directly from the cache similar to the RSC guidance of fetching directly in the component where the data is used.  Mostly trying to prevent people seeing this approach and then unnecessarily using it with client-initiated fetches.

* **Double unwrap.** Reading the value requires two calls: `use(context)` to get the Promise, then `use(promise)` to get its resolved value. The context value itself is not awaited.

> * **Revalidation scope.** Because the Promise is created in the component that provides it, revalidating the data requires re-rendering that component. Hoisting data into a component high in the tree means everything below it re-renders to update the data.

Reviewer 2: I wonder if this is a problem with "[lift content up](https://overreacted.io/before-you-memo/)" and the React compiler.  Even if you pass a promise to a provider rendered in a server component, it only affects the consumers that would read from the associated context.

Author: I think this was a different thing, and maybe more specific to particular frameworks. 

The issue I was flagging with this pattern is that if you have nested routes like:

```
- users
 - page.js  <- context here
 - [id] 
   - page.js <- read here
```

then you would more often need to call `revalidate('/users')` instead of the more scoped `revalidate('/users/[id]')`, since the data the lower page.js is reading from is hoisted higher in the tree and only re-fetched when the higher page.js is refetched.

Author: * **Refetching scope for RSCs**: When this pattern is used with Server Components, refetching the promise requires refetching the server component that sets the promise in context. Avoid setting the promise in context high in the tree, since this would require refetching large parts of the app unnecessarily.

> * **One Promise per context.** Each piece of data needs its own context. Reaching for more contexts to share more data is usually a sign that the data should be fetched closer to where it's used.

Reviewer 2: The "One Promise per context" feels a bit more prescriptive than usual.  I wonder if linking to `/learn/passing-data-deeply-with-context#before-you-use-context` would be preferable here.

Author: +1

</Pitfall>

---

## Usage (Promises) {/*usage-promises*/}

### Reading a Promise with `use` {/*reading-a-promise-with-use*/}

Call `use` with a Promise to read its resolved value. The component will [suspend](/reference/react/Suspense) while the Promise is pending.

```js [[1, 4, "use(albumsPromise)"]]
import { use } from 'react';

function Albums({ albumsPromise }) {
  const albums = use(albumsPromise);
  return (
    <ul>
      {albums.map(album => (
        <li key={album.id}>
          {album.title} ({album.year})
        </li>
      ))}
    </ul>
  );
}
```

> The component that calls <CodeStep step={1}>`use`</CodeStep> must be wrapped in a [`Suspense`](/reference/react/Suspense) boundary. While the Promise is pending, the Suspense fallback is displayed. Once the Promise resolves, React reads the value with `use` and replaces the fallback with the rendered component.

Reviewer 2: > The component that calls use must be wrapped in a [Suspense](http://localhost:3000/reference/react/Suspense) boundary

This uses `must` as if an error would occur, but I deleted Suspense and it still worked.  It wasn't pretty but there wasn't any error either.

Reviewer 4: https://github.com/reactjs/react.dev/pull/8305/changes/c46d999a1247842c2d1a581dd586159be00a8df6
Fixed here, thanks!

    id: 10,
    title: 'The Beatles',
    year: 1968
  }];
}
```

</Sandpack>

<Solution />

</Recipes>

<Pitfall>

##### Promises passed to `use` must be cached {/*promises-must-cached*/}

Promises created during render are recreated on every render, which causes React to show the Suspense fallback repeatedly and prevents content from appearing.

```js
function Albums() {
  // 🔴 `fetch` creates a new Promise on every render.
  const albums = use(fetch('/albums'));
  // ...
}
```

Instead, pass a Promise from a cache, a Suspense-enabled framework, or a Server Component:

```js
// ✅ fetchData reads the Promise from a cache.
const albums = use(fetchData('/albums'));
```

</Pitfall>

<DeepDive>

#### Why are Promises recreated on every render? {/*why-promises-recreated*/}

[React doesn't preserve state for renders that suspended before mounting](/reference/react/Suspense#caveats). After each suspension, React retries rendering from scratch, so any Promise created during render is recreated.

Common ways a Promise can be unintentionally recreated during render:

```js
function Albums() {
  // 🔴 `fetch` creates a new Promise on every render.
  const albums = use(fetch('/albums'));

>   // 🔴 Inline `async` function calls create a new Promise on every render.

Reviewer 2: **nit**: I don't think it's the fact of the call being inline (whether inline is interpreted as the function is defined inline, or invoked inline as the argument to use).  It's that the function returns a promise which isn't cached.

Maybe "Uncached `async` function calls create a new Promise on every render."?

Reviewer 4: Fixed! https://github.com/reactjs/react.dev/pull/8305/changes/c46d999a1247842c2d1a581dd586159be00a8df6 thanks

</Sandpack>

> ### Reading a Promise from context {/*reading-a-promise-from-context*/}

Reviewer 3: Great thinking on adding this section!

> You can pass a Promise through context to share data without prop drilling. Pass the Promise as the context value, then call `use(context)` to read the Promise and `use(promise)` to read its resolved value:

Author: > You can pass a Promise through context to share data without prop drilling. 

Hm, this makes it sound like you have to pass a promise to share data, but you can also just await the promise and put the response in context. When should you await vs no?

Reviewer 4: Tried to address here: https://github.com/reactjs/react.dev/pull/8305/commits/e7be13d2c36ed066fcf93b242df2ec7f908140c9

```js
import { use } from 'react';
import { UserContext } from './UserContext';

function Profile() {
  const userPromise = use(UserContext);
  const user = use(userPromise);
  return <h1>{user.name}</h1>;
}
```

Wrap the components that read the Promise in [`<Suspense>`](/reference/react/Suspense) so only that subtree suspends while the Promise is pending. See [Usage (Promises)](#usage-promises) below for more on reading Promises with `use`.

> <Pitfall>

Author: These are more caveats than pitfalls.

Reviewer 4: Adressed: https://github.com/reactjs/react.dev/pull/8305/commits/e7be13d2c36ed066fcf93b242df2ec7f908140c9

Passing a Promise through context is convenient, but comes with trade-offs. Prefer passing the Promise as a [prop](#reading-a-promise-with-use) from a [Server Component](/reference/rsc/server-components) when possible.

> * **Double unwrap.** Reading the value requires two calls: `use(context)` to get the Promise, then `use(promise)` to get its resolved value. The context value itself is not awaited.

Author: This seems framed as a problem (for example, maybe i assume this means reading multiple contexts are slow). But it's really just something to not forget to do.

Reviewer 4: Tried to adress it here: https://github.com/reactjs/react.dev/pull/8305/commits/e7be13d2c36ed066fcf93b242df2ec7f908140c9

> * **Revalidation scope.** Because the Promise is created in the component that provides it, revalidating the data requires re-rendering that component. Hoisting data into a component high in the tree means everything below it re-renders to update the data.

Author: Also React/RSCs don't really have "revalidation" mentioned anywhere else in the docs. Other frameworks might not even have it or call it "revalidation".

Reviewer 4: Updated: https://github.com/reactjs/react.dev/pull/8305/commits/e7be13d2c36ed066fcf93b242df2ec7f908140c9

</DeepDive>

---

### Displaying an error with an Error Boundary {/*displaying-an-error-with-an-error-boundary*/}

If the Promise passed to `use` is rejected, the error propagates to the nearest [Error Boundary](/reference/react/Component#catching-rendering-errors-with-an-error-boundary). Wrap the component that calls `use` in an Error Boundary to display a fallback when the Promise is rejected.

<Sandpack>

```js src/App.js active
import { use, Suspense, useState, useTransition } from "react";
import { ErrorBoundary } from "react-error-boundary";
import { fetchData } from "./data.js";

export default function App() {
  const [albumsPromise, setAlbumsPromise] = useState(
    () => fetchData('/the-beatles/albums')
  );
  const [isPending, startTransition] = useTransition();

  function handleRefresh() {
    startTransition(() => {

>       setAlbumsPromise(fetchData('/the-beatles/albums'));

Reviewer 4: fixed https://github.com/reactjs/react.dev/pull/8305/changes/ceab6dadbfd754ef3c83a02b8a3cd182d501ed65

    });
  }

  return (
    <>
      <button onClick={handleRefresh} disabled={isPending}>
        {isPending ? 'Refreshing...' : 'Refresh'}
      </button>

>       <ErrorBoundary fallback={<p>⚠️Something went wrong</p>}>

Reviewer 4: fixed https://github.com/reactjs/react.dev/pull/8305/changes/ceab6dadbfd754ef3c83a02b8a3cd182d501ed65

</Sandpack>

<DeepDive>

#### Should I resolve a Promise in a Server or Client Component? {/*resolve-promise-in-server-or-client-component*/}

A Promise can be passed from a Server Component to a Client Component and resolved in the Client Component with the `use` API. You can also resolve the Promise in a Server Component with `await` and pass the resolved value to the Client Component as a prop. The difference comes down to how much of your UI you want to reveal before the data is ready.

Using `await` in a [Server Component](/reference/rsc/server-components) suspends the Server Component itself, so React won't continue rendering it until the Promise resolves:

```js
// Server Component
export default async function App() {
  const messageContent = await fetchMessage();
  return <Message messageContent={messageContent} />;
}
```

Passing the Promise to a Client Component doesn't suspend the Server Component. The Server Component returns immediately, and the Client Component suspends when it calls `use`:

```js
// Server Component
export default function App() {
  const messagePromise = fetchMessage();
  return <Message messagePromise={messagePromise} />;
}
```

> In both cases, the component that reads the Promise suspends. Make sure there's a [`<Suspense>`](/reference/react/Suspense) boundary above it so React can display a fallback there and keep rendering the rest of the page. Without a boundary above the suspending component, the suspension bubbles up the tree and blocks the surrounding UI until the Promise resolves.

Reviewer 2: Nice improvement 👍

Passing the Promise down is useful when a Client Component needs the data and you want to reveal as much of the surrounding UI as possible while the Promise is pending. Because the Client Component is the component that suspends, you can place the `<Suspense>` boundary close to it and let the rest of the page render right away. See [Displaying a fallback while content is loading](/reference/react/Suspense#displaying-a-fallback-while-content-is-loading) for more on boundary placement.

> When the surrounding UI is small or depends on the same data, prefer `await` on the server. If a Server Component above already awaits the data, pass the resolved value down instead of creating a new Promise to call `use`.

Reviewer 2: One thing that's not clear to me here is _why_ you should prefer `await` on the server in this case.  Maybe just stating the advantage would be better?

One thing I thought that was different between these approaches was the time it takes for the promise to reach the client from the server.  If you suspend on the server, you'd start streaming the fallback, however if suspend on the client and your async call is fast enough, by the time your promise reaches the client with the rest of the response, it could be resolved already and so you may not show the user an unnecessary fallback.  I could be completely wrong here.

Reviewer 4: Does this make it clearer? I'm still not sure about this section! https://github.com/reactjs/react.dev/pull/8305/changes/c46d999a1247842c2d1a581dd586159be00a8df6

Reviewer 2: I think it's clearer and the added code block and comment with `// Will suspend until the data is available.` was a nice addition.  

Long wall of text incoming.

Thought it's clearer it's still a bit difficult to parse the recommendations.  I think I'm seeing it as there are 2 different things.  1) what the actual behavioral differences are and 2) what the implications of those behavioral differences are which may lead to picking one approach over the other.  I'm not fully sure of 1) myself so 2) isn't clear.  Ideally I'd hope the section could clarify both 1) and 2).

In terms me trying to make sense of the actual behavior differences, I see 4 options.  a) await in RSC without Suspense, b) use() in Client Component without Suspense c) await in an RSC with Suspense d), use() in a Client Component  with Suspense.  

> In both cases, wrap the component that reads the Promise in a Suspense boundary so React can display a fallback while the Promise is pending.

I think the code examples help clarify the behavior of  a) and b), but in the end we recommend to always choose either c) or d), but the difference between c) and d) isn't made clear.  If I'm using Suspense, what does it matter if we use `use()` in a Client Component or `await` in a Server Component?  Both unblock the initial server render (at least in NextJS).  I think there is some difference between c) and d) (both in NextJS and framework-independent), however subtle, and that's what would be useful to clarify IMO.

One structure that might help is: 
- first start clarifying that both use() and await() without Suspense , will suspend, like you already do, but do so without making recommendations yet (because later down you mention to always use Suspense).
- Then show code examples with Suspense and clarify the behavioral differences of use() vs await with Suspense. 
- Based on those behavioral differences, provide recommendations.

This is just how it would make sense to me, but I also get it would make a pretty long deep dive as well.

One final nitpick about this part is that " In both cases, wrap the component that reads the Promise in a Suspense boundary" can be interpreted literally and lead to loading states that are too fine-grained (spinners everywhere) instead of [revealing things together at once](https://react.dev/reference/react/Suspense#revealing-content-together-at-once).

> In general, prefer resolving the Promise with await in a Server Component. The Server Component suspends until the Promise resolves, and the Client Component receives the resolved value as a prop

I still can't tell why this would be recommended here.  It seems that the only advantage is you'd be passing the data instead of a promise with the data.  IMHO mentioning the behavior differences/tradeoffs may be enough if there's no clear winner. 

> Resolve the Promise with use in a Client Component for data that isn't important for the surrounding UI to render, like below-the-fold content, or content revealed by a hover or click (popovers, tooltips). The Server Component starts the Promise but doesn't await it, so it returns immediately.

What makes `use()` well-suited for this case?  For below the fold content you can use Suspense with await in an RSC as well.  Maybe you are referring to client-initiated fetches with `use()` here?  I do think server-initiated fetches can be too eager for things that are hidden that users may never interact with.  If it were that, maybe it would warrant a change in the title of the deep dive since it's not just where to resolve a Promise (and assume it's initiated on the server) but where to create the Promise as well.

> The Server Component starts the Promise but doesn't await it, so it returns immediately. The Client Component suspends when it calls use:
> 
> ...
> This way the Client Component suspends without blocking the rest of the page from rendering.

Without Suspense (the associated code example doesn't use Suspense) it does block the rest of the page.  This is what you mentioned in [this thread](https://github.com/reactjs/react.dev/pull/8305#discussion_r2814590686) and what I was testing in NextJS. Not sure if it is framework-specific though.

I can comment directly on the commit next time if you prefer, just though trying to keep a thread going with everything might be easier.

Reviewer 4: Tried to adress it better: https://github.com/reactjs/react.dev/pull/8305/commits/e7be13d2c36ed066fcf93b242df2ec7f908140c9

}
```

Prefer `await` in a Server Component when possible, since it keeps the data fetching on the server. If a Server Component above already awaits the data, pass the resolved value down as a prop instead of creating a new Promise to call `use`.

> Use `use` in a Client Component to push the data read deeper in the tree, so more of the surrounding UI can render while the Promise is pending. A common case is interactive content like popovers and tooltips, where the data is needed only after a hover or click. Client Components can't `await`, so they rely on `use` to suspend on a Promise.

Author: ```suggestion
You can also pass promise as a prop to a Client Component without awaiting it, and then read it with `use(promise)` to suspend deeper in the tree. This allows more of the surrounding UI to complete while the Promise is pending. A common case is interactive content like popovers and tooltips, where the data is needed only after a hover or click. Client Components can't `await`, so they rely on `use` to suspend on a Promise.
```

```
</Sandpack>

---

---

## Troubleshooting {/*troubleshooting*/}

### I'm getting an error: "Suspense Exception: This is not a real error!" {/*suspense-exception-error*/}

You are calling `use` inside a try-catch block. `use` throws internally to integrate with Suspense, so it cannot be wrapped in try-catch. Instead, wrap the component that calls `use` in an [Error Boundary](#displaying-an-error-with-an-error-boundary) to handle errors.

```jsx
function Albums({ albumsPromise }) {
  try {

>     // ❌ Don't wrap `use` in try-catch

Author: No but it's one of those things where when you hit it, people get mad that it's not in the docs

```
</Sandpack>

---

> ---

Reviewer 3: This is an extra line that I do not think is needed

Reviewer 4: Thank you! fixed in https://github.com/reactjs/react.dev/pull/8305/changes/92208c21624e244d5608a27586def0b6c8cbe6a6
